### PR TITLE
feat: support automatic sharding

### PR DESCRIPTION
**Pull request by Duologic**
Extend this library to run KSM in automatic sharding mode, open sourcing this config as used within Grafana Labs.

### CODE REVIEW

File: kube-state-metrics/main.libsonnet [hunk 1]
```diff
@@ -57,7 +77,11 @@ local d = import 'github.com/jsonnet-libs/docsonnet/doc-util/main.libsonnet';
             self.rbac.service_account.metadata.name
           )
           + deployment.spec.template.spec.securityContext.withRunAsUser(65534)
-          + deployment.spec.template.spec.securityContext.withRunAsGroup(65534),
+          + deployment.spec.template.spec.securityContext.withRunAsGroup(65534)
+          + deployment.configMapVolumeMount(self.config_map, self.config_path),
+
+        // Add hidden statefulset to allow mixins on both
+        statefulset:: {},
```
Comment (logyball) [inline kube-state-metrics/main.libsonnet:84]: This doesn't seem to work for me.  When defining a KSM file like this:

```jsonnet

... 

    meta.environment.baseEnvironment(
      root.app,
      cluster,
      root.namespace,
      data={
        kube_state_metrics:
          local kubeStateMetricsVersion = (import './waves.libsonnet')[cluster.wave];
          ksm.new(namespace=root.namespace)
          + ksm.withAutomaticSharding(10)
          + ksm.withPriorityClass('critical'),
      }
    ),

  envs: meta.environment.generateEnvs(root.clusters, root.environment),
}

...

```

I get the following output:

```sh
❯ tk eval ksonnet/environments/kube-state-metrics-2/main.jsonnet -e <environment> | jq '.data.kube_state_metrics | keys'
[
  "config_map",
  "pdb",
  "rbac",
]
```

Meaning the STS just straight up doesn't show up.

I can rectify this by setting the STS in line 84 to a visible field:

```
statefulset: {}
```

and modifying the withPriorityClass function slightly:

```jsonnet
withPriorityClass(priorityClassName):: {
    deployment+:
      k.apps.v1.deployment.spec.template.spec.withPriorityClassName(priorityClassName),
    statefulset+: if super.statefulset == {} then {} else k.apps.v1.statefulSet.spec.template.spec.withPriorityClassName(priorityClassName),
  },
```

WDYT?

**Review by logyball — CHANGES_REQUESTED** [review]
Comment regarding the statefulset, it doesn't seem to get rendered into json and/or YAML as a k8s manifest as it stands

File: kube-state-metrics/main.libsonnet (hunk 1 above)
Comment (Duologic) [reply kube-state-metrics/main.libsonnet:84]: I've done it differently, exposing the statefulset with `:::` when invoking `withAutomaticSharding`.

Comment (logyball) [reply kube-state-metrics/main.libsonnet:84]: TIL about `:::`

**Review by logyball — APPROVED** [review]
LGTM! Nice work on this

**Pull request merged**